File: HelloBrowser3D/HelloBrowser3D/ViewController.m
//
//  ViewController.m
//
//  Hello3DBrowser
//
//  Copyright (c) 2015 Aumentia. All rights reserved.
//
//  Written by Pablo GM <info@aumentia.com>, January 2015
//

#import "ViewController.h"
#import "CaptureSessionManager.h"

#define kAccelerometerFrequency        90.0     //Hz

@interface ViewController ()
{
    // Browser instance
    _DBrowser                   *my3dBrowser;
    
    // Gestures
    UIPanGestureRecognizer      *_panRecognizer;
    UIPinchGestureRecognizer    *_pinchRecognizer;
    UIRotationGestureRecognizer *_rotationRecognizer;
    
    // Camera
    CaptureSessionManager       *_captureManager;
    UIView                      *_cameraView;
    
    // Radar Plugin
    PoiRadarController          *_radarController;
    NSMutableArray              *_poisArray;
    
    // Motion
    CMMotionManager             *motionManager;
    
    Poi                         *currentSelectedPOI;
    
    BOOL                        isIphone;
    
    CLLocation                  *lastLocation;
}

@end

@implementation ViewController

@synthesize locationManager;


#pragma mark - View Life Cycle

- (void)viewWillAppear:(BOOL)animated
{
    [super viewWillAppear:YES];
	// Do any additional setup after loading the view, typically from a nib.
    
    if (UI_USER_INTERFACE_IDIOM() == UIUserInterfaceIdiomPhone)
        isIphone = YES;
    else
        isIphone = NO;
    
    // Add camera
    [self initCapture];
    
    // Initiliaze
    my3dBrowser = [[_DBrowser alloc] initBrowser:CGRectMake(0, 0, self.view.bounds.size.width, self.view.bounds.size.height) withKey:@"627a859b970b7d2cbb337e99ada9610f13ae1b64" setDebug:YES representation:geographical];
    
    // Delegate
    [my3dBrowser setDelegate:self];
    
    // Transparent brackground
    [my3dBrowser setBackgroundColor:[UIColor clearColor]];
    
    // Add view
    [self.view addSubview:my3dBrowser];
    [self.view bringSubviewToFront:my3dBrowser];
    
    // Add gestures
    [self addGestureEvents];
    
    // Init location manager
    [self setupLocationManager];
    
    // Set connection properties
    //[my3dBrowser setUrlConnections:ASYNC displayLoading:YES];
    
    // Init motion
    [self initMotionManager];
    [my3dBrowser setMotionManager:motionManager];
    
    // Radar
    _poisArray = [[NSMutableArray alloc] init];
    [self setupRadar];
    
    self.view.backgroundColor = [UIColor clearColor];
}

- (void)viewWillDisappear:(BOOL)animated
{
    [super viewWillDisappear:animated];
    
    // Remove POIs
    [my3dBrowser removeAllPois];
    
    // Remove camera
    [self removeCapture];
    
    // Remove gestures
    [self removeGestures];
    
    // Remove motion manager
    [self removeMotionManager];
    
    // Remove location manager
    [self removeLocationManager];
    
    // Remove radar
    [self removeRadar];
    
    // Remove notifications
    [[NSNotificationCenter defaultCenter] removeObserver:self];
    
    // Release memory
    [my3dBrowser releaseBrowser];
    [my3dBrowser removeFromSuperview];
    [self releaseMemory];
    
    for (UIView *view in [self.view subviews])
    {
        [view removeFromSuperview];
    }
    
    [[UIDevice currentDevice] endGeneratingDeviceOrientationNotifications];

    [[NSNotificationCenter defaultCenter] removeObserver:self];
}


#pragma mark - Add POIs

- (BOOL)addPOIs:(CLLocation*)location
{
    // SAMPLE COORDS
    float _lat = location.coordinate.latitude;
    float _lon = location.coordinate.longitude;
    
    // Add a set of 2D and 3D POIs
    if( [self addPOI:[[CLLocation alloc]initWithLatitude:_lat - 1 longitude:_lon + 1]] &&
       
       [self addPOI:[[CLLocation alloc]initWithLatitude:_lat + 1 longitude:_lon + 1]])
        
    {
        if ( [self addPOI2D:[[CLLocation alloc]initWithLatitude:_lat + 1 longitude:_lon - 1]] &&
             [self addPOI2D:[[CLLocation alloc]initWithLatitude:_lat + 1 longitude:_lon - 2]] )
        {
            return YES;
        }
    }
    
    return NO;
    
}

- (BOOL)addPOI2D:(CLLocation*)location
{
    static int _counter2D = 0;
    
    //// CONFIGURE 2D POIS
    Poi2D* poi = [[Poi2D alloc] init];
    
    poi.altitude = 0;
    
    poi.selectedAlpha = 1;
    
    poi.notSelectedAlpha = .7;
    
    poi.poiSize = 5;
    
    [poi setLocation:location];
    
    switch (_counter2D) {
        case 0:
        {
            [poi setTextureSelectionPath:@"Viking_alpha.png"];
            [poi setTextureNonSelectionPath:@"Transformer_alpha.png"];
            break;
        }
        case 1:
        {
            [poi setTextureSelectionPath:@"Zombie.png"];
            [poi setTextureNonSelectionPath:@"Tv_alpha.png"];
            break;
        }
        default:
            break;
    }
    
    [_poisArray addObject:poi];
    
    // Add POI and get its unique id
    NSInteger uIdPOI = [my3dBrowser add:poi];
    
    _counter2D++;
    
    if (_counter2D == 2)
        _counter2D = 0;
    
    return uIdPOI!= -1 ? YES : NO;
}

- (BOOL)addPOI:(CLLocation*)location
{
    static int _counter = 0;
    
    Poi3D* poi = [[Poi3D alloc] init];
    
    poi.playAnimationOnStart = YES;
    
    poi.altitude = 0;
    
    poi.poiSize = 2;
    
    poi.poiCam =  POINTING_CAMERA;
    
    [poi setLocation:location];
    
    [_poisArray addObject:poi];
    
    if(_counter % 2 == 0)
    {
        poi.pathEAD = @"data/Models/pois/knight/Dark_knight.EAD";
        [poi setPoiSize:3];
    }
    else
    {
        poi.pathEAD = @"data/Models/pois/zombie/Zombie_head.EAD";
        [poi setPoiSize:10];
    }
    
    NSInteger uIdPOI = [my3dBrowser add:poi];
    
    _counter++;
    
    if (_counter == 2)
        _counter = 0;
    
    return uIdPOI == -1 ? NO : YES;
}


#pragma mark - Gesture Management

- (void)addGestureEvents
{
    // PAN
    if(_panRecognizer == nil)
    {
        _panRecognizer = [[UIPanGestureRecognizer alloc] initWithTarget:self action:@selector(move:)];
        [_panRecognizer setMinimumNumberOfTouches:1];
        [_panRecognizer setMaximumNumberOfTouches:1];
        [_panRecognizer setDelegate:self];
        [self.view addGestureRecognizer:_panRecognizer];
    }
    
    // PINCH
    if (_pinchRecognizer == nil)
    {
        _pinchRecognizer = [[UIPinchGestureRecognizer alloc] initWithTarget:self action:@selector(zoomPinch:)];
        [self.view addGestureRecognizer:_pinchRecognizer];
    }
    
    // ROTATION
    if(_rotationRecognizer)
    {
        _rotationRecognizer = [[UIRotationGestureRecognizer alloc] initWithTarget:self action:@selector(spin:)];
        [self.view addGestureRecognizer:_rotationRecognizer];
    }
    
}

- (void)spin:(UIRotationGestureRecognizer *)gestureRecognizer {
    
    //CGFloat rotation = gestureRecognizer.rotation;
    //NSLog(@"Rotation: %f", rotation);
    
}

- (void)zoomPinch:(UIPinchGestureRecognizer *)recognizer
{
    //NSLog(@"Pinch scale: %f, velocity: %f", recognizer.scale, recognizer.velocity);
}

-(void)move:(UIPanGestureRecognizer *)recognizer
{
    [my3dBrowser spin3DPOI:recognizer];
}

- (void)removeGestures
{
    NSArray* gesturesArr = [self.view gestureRecognizers];
    
    if ([gesturesArr count] > 0)
    {
        [self.view removeGestureRecognizer:_panRecognizer];
        [self.view removeGestureRecognizer:_pinchRecognizer];
        [self.view removeGestureRecognizer:_rotationRecognizer];
    }
}


#pragma mark - Memory Management

- (void)didReceiveMemoryWarning
{
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

- (void)releaseMemory
{
    locationManager             = nil;
    my3dBrowser                 = nil;
    my3dBrowser.delegate        = nil;
    
    currentSelectedPOI          = nil;
    
    // GESTURES
    _pinchRecognizer            = nil;
    _panRecognizer              = nil;
    _rotationRecognizer         = nil;
    
    // RADAR
    _poisArray                  = nil;
    _radarController            = nil;
}


#pragma mark - Motion Manager

- (void)initMotionManager
{
    if (motionManager == nil)
    {
        // Set up MotionManager
        motionManager = [[CMMotionManager alloc] init]; // motionManager is an instance variable
        
        motionManager.showsDeviceMovementDisplay    = YES; //calibration popup
        
        motionManager.accelerometerUpdateInterval   = 1.0 / kAccelerometerFrequency; // 100Hz
        motionManager.deviceMotionUpdateInterval    = 1.0 / kAccelerometerFrequency;
        
        [motionManager startAccelerometerUpdates];
        [motionManager startDeviceMotionUpdatesUsingReferenceFrame:CMAttitudeReferenceFrameXMagneticNorthZVertical];
    }
}

- (void)removeMotionManager
{
    if (motionManager != nil)
    {
        [motionManager stopAccelerometerUpdates];
        [motionManager stopDeviceMotionUpdates];
        motionManager = nil;
    }
}


#pragma mark - Location Manager

- (void)setupLocationManager
{
    // LOCATION
    if (locationManager == nil)
    {
		locationManager = [[CLLocationManager alloc] init];
        locationManager.headingFilter = kCLHeadingFilterNone;
		locationManager.desiredAccuracy = kCLLocationAccuracyBest;
		locationManager.delegate = self;
        
        if ( [locationManager respondsToSelector:@selector(requestAlwaysAuthorization)] )
        {
            [locationManager requestAlwaysAuthorization];
        }
        
		[locationManager startUpdatingLocation];
        [locationManager startUpdatingHeading];
	}
}

-(void)removeLocationManager
{
    if (locationManager != nil)
    {
        [locationManager stopUpdatingHeading];
        locationManager = nil;
        lastLocation = nil;
    }
}


#pragma mark - Location Delegate

- (void)locationManager:(CLLocationManager *)manager didFailWithError:(NSError *)error
{
    NSLog(@"location error");
}

- (void)locationManager:(CLLocationManager *)manager didUpdateToLocation:(CLLocation *)newLocation fromLocation:(CLLocation *)oldLocation
{
    CLLocationDistance distanceInMeters = [newLocation distanceFromLocation:oldLocation];
    
    if(my3dBrowser != nil && distanceInMeters >= 0)
    {
        if (lastLocation == nil)
        {
            // Add POIs
            [self addPOIs:newLocation];
            
            [self refreshPois];
        }
        lastLocation = newLocation;
        
        NSLog(@"location updated lon %f and lat %f", newLocation.coordinate.longitude, newLocation.coordinate.latitude);
        
        // ARBrowser
        [my3dBrowser updateLocation:newLocation];
        
        // Radar
        [self setUserLocationInRadar:newLocation];
    }
}

- (void)locationManager:(CLLocationManager *)manager didUpdateHeading:(CLHeading *)newHeading
{
    [_radarController setHeading:newHeading];
}


#pragma mark - Render Delegate

- (void)poiClicked:(Poi *)poi
{
    NSLog(@"Poi %ld clicked!!!", (long)poi.uId);
    
    // Let's refresh the texture on click
    if ( [poi isKindOfClass:[Poi2D class]])
    {
        [(Poi2D*)poi setTextureSelectionPath:@"Knight_fullbody.png"];
        
        [my3dBrowser refreshPoiTextures:(Poi2D*)poi];
    }
    
    currentSelectedPOI = poi;
}

- (void)poiSelected:(Poi *)poi
{
    NSLog(@"Poi %ld selected!!!", (long)poi.uId);
    
    currentSelectedPOI = poi;
}

- (void)poisInView:(NSMutableArray *)poisArray
{
    //NSLog(@"Pois in view %d", [poisArray count]);
    [_radarController setSelectedPoi:currentSelectedPOI];
    [_radarController refresh];
}


#pragma mark - Camera management

- (void)initCapture {
	
    // init capture manager
    _captureManager = [[CaptureSessionManager alloc] init];
    
    // set video streaming quality
    _captureManager.captureSession.sessionPreset = AVCaptureSessionPresetMedium;   //480x360
    
    [_captureManager setOutPutSetting:[NSNumber numberWithInt:kCVPixelFormatType_32BGRA]]; //kCVPixelFormatType_420YpCbCr8BiPlanarVideoRange
    
    [_captureManager addVideoInput:AVCaptureDevicePositionBack]; //AVCaptureDevicePositionFront / AVCaptureDevicePositionBack
    [_captureManager addVideoOutput];
    [_captureManager addVideoPreviewLayer];
    
    CGRect layerRect = self.view.bounds;
    
    [[_captureManager previewLayer] setOpaque: 0];
    [[_captureManager previewLayer] setBounds:layerRect ];
    [[_captureManager previewLayer] setPosition:CGPointMake( CGRectGetMidX(layerRect), CGRectGetMidY(layerRect) ) ];
    
    // create a view, on which we attach the AV Preview layer
    _cameraView = [[UIView alloc] initWithFrame:self.view.bounds];
    [[_cameraView layer] addSublayer:[_captureManager previewLayer]];
    
    // add the view we just created as a subview to the View Controller's view
    [self.view addSubview: _cameraView];
    
    // start !
    [self performSelectorInBackground:@selector(start_captureManager) withObject:nil];
    
}

- (void)removeCapture
{
    [_captureManager.captureSession stopRunning];
    [_cameraView removeFromSuperview];
    _captureManager     = nil;
    _cameraView         = nil;
}

- (void)start_captureManager
{
    @autoreleasepool
    {
        [[_captureManager captureSession] startRunning];
    }
}


#pragma mark - Radar

- (void)setupRadar
{
    if (_radarController == nil)
    {
        _radarController = [[PoiRadarController alloc] initWithKey:@"cffd9090c2d30af63325a5e4faeb795edd8bf12d"];
        
        UIImage* radarImage;
        
        if(UI_USER_INTERFACE_IDIOM() == UIUserInterfaceIdiomPhone)
        {
            radarImage = [UIImage imageNamed:@"resources.bundle/iPhoneRadar.png"];
        }else{
            radarImage = [UIImage imageNamed:@"resources.bundle/iPadRadar.png"];
        }
        
        [_radarController.view setFrame:CGRectMake(self.view.frame.origin.x + 10, self.view.frame.origin.y + 10, radarImage.size.width, radarImage.size.height)];
        
        [_radarController setRadarImage:radarImage];
        
        [_radarController setPoiImage:[UIImage imageNamed:@"resources.bundle/poi.png"]];
        [_radarController setPoiImageSelected:[UIImage imageNamed:@"resources.bundle/poiSelected.png"]];
        [_radarController setNorthImage:[UIImage imageNamed:@"resources.bundle/northRadar.png"]];
        
        [self.view addSubview:_radarController.view];
    }
}

- (void)removeRadar
{
    [_radarController.view removeFromSuperview];
    _radarController    = nil;
}

-(void)rotateRadarView:(double)degrees
{
    [_radarController rotateRadarView:degrees];
}

-(void) refreshPois
{
    @synchronized(self)
    {
        [_radarController setPoisArray:_poisArray];
        [_radarController reset];
        [_radarController refresh];
    }
}

-(void)setUserLocationInRadar:(CLLocation *)_userLocation
{
    [_radarController setUserLocation:_userLocation];
}


#pragma mark - Autorotate

- (BOOL)shouldAutorotate
{
    return NO;
}

#pragma mark - Util functions

-(UIColor*)colorWithHexString:(NSString*)hex
{
    NSString *cString = [[hex stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceAndNewlineCharacterSet]] uppercaseString];
    
    // String should be 6 or 8 characters
    if ([cString length] < 6) return [UIColor grayColor];
    
    // strip 0X if it appears
    if ([cString hasPrefix:@"0X"]) cString = [cString substringFromIndex:2];
    
    if ([cString length] != 6) return  [UIColor grayColor];
    
    // Separate into r, g, b substrings
    NSRange range;
    range.location = 0;
    range.length = 2;
    NSString *rString = [cString substringWithRange:range];
    
    range.location = 2;
    NSString *gString = [cString substringWithRange:range];
    
    range.location = 4;
    NSString *bString = [cString substringWithRange:range];
    
    // Scan values
    unsigned int r, g, b;
    [[NSScanner scannerWithString:rString] scanHexInt:&r];
    [[NSScanner scannerWithString:gString] scanHexInt:&g];
    [[NSScanner scannerWithString:bString] scanHexInt:&b];
    
    return [UIColor colorWithRed:((float) r / 255.0f)
                           green:((float) g / 255.0f)
                            blue:((float) b / 255.0f)
                           alpha:1.0f];
}

@end

File: HelloBrowser3D/HelloBrowser3DSwift/HelloBrowser3DSwift/ViewController.swift
//
//  ViewController.swift
//  HelloBrowser3DSwift
//
//  Created by Pablo GM on 11/09/15.
//  Copyright © 2015 Aumentia Technologies SL. All rights reserved.
//

let kAccelerometerFrequency = 90.0     //Hz

import UIKit

class ViewController: UIViewController, EDBrowserProtocol
{
    // Location manager instance
    var _locationManager:CLLocationManager!
    
    // Browser instance
    var _my3dBrowser:_DBrowser!
    
    // Gestures
    var _panRecognizer:UIPanGestureRecognizer!
    var _pinchRecognizer:UIPinchGestureRecognizer!
    var _rotationRecognizer:UIRotationGestureRecognizer!
    
    // Camera
    var _captureManager:CaptureSessionManager!
    var _cameraView:UIView!
    
    // Radar Plugin
    var _radarController:PoiRadarController!
    var _poisArray:NSMutableArray!
    
    // Motion
    var _motionManager:CMMotionManager!
    
    var _currentSelectedPOI:Poi!
    
    var _lastLocation:CLLocation!
    
    

    // MARK: - View Life Cycle
    
    override func viewDidLoad()
    {
        super.viewDidLoad()
        // Do any additional setup after loading the view, typically from a nib.
    }
    
    override func viewWillAppear(_ animated: Bool)
    {
        super.viewWillAppear(animated)
        
        // Add camera
        initCapture()
        
        // Initialize
        _my3dBrowser = _DBrowser(browser: self.view.bounds, withKey: "7b62a465ea1965171a81aed536b313c62e290478", setDebug: false, representation: poiRepresentation.geographical)
        
        assert(_my3dBrowser != nil, "Check API Key")
        
        // Delegate
        _my3dBrowser.delegate = self
        
        // Transparent background
        _my3dBrowser.backgroundColor = UIColor.clear
        
        // Add view
        self.view.addSubview(_my3dBrowser)
        self.view.bringSubview(toFront: _my3dBrowser)
        
        // Add gestures
        addGestureEvents()
        
        // Init location manager
        setupLocationManager()
        
        // Init motion
        initMotionManager()
        _my3dBrowser.setMotionManager(_motionManager)
        
        // Radar
        _poisArray = NSMutableArray()
        setupRadar()
        
        self.view.backgroundColor = UIColor.clear
    }
    
    override func viewWillDisappear(_ animated: Bool)
    {
        super.viewWillDisappear(animated)
        
        // Remove POIs
        _my3dBrowser.removeAllPois()
        
        // Remove camera
        removeCapture()
        
        // Remove gestures
        removeGestures()
        
        // Remove motion manager
        removeMotionManager()
        
        // Remove location manager
        removeLocationManager()
        
        // Remove radar
        removeRadar()
        
        // Release memory
        _my3dBrowser.release()
        _my3dBrowser.removeFromSuperview()
        releaseMemory()
    }

    
    // MARK: - Memory Management
    
    override func didReceiveMemoryWarning()
    {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }

    func releaseMemory()
    {
        _locationManager        = nil
        _my3dBrowser            = nil
        _my3dBrowser.delegate   = nil
        
        _currentSelectedPOI     = nil
        
        // GESTURES
        _pinchRecognizer        = nil
        _panRecognizer          = nil
        _rotationRecognizer     = nil
        
        // RADAR
        _poisArray.removeAllObjects()
        _poisArray              = nil
        _radarController        = nil
    }
}


// MARK: - 3D Browser
extension ViewController
{
    // MARK: - Add POIs
    
    func addPoi(location: CLLocation) -> Bool
    {
        struct Counter { static var _counter = 0 }
        
        let poi:Poi3D =  Poi3D()
        
        poi.playAnimationOnStart = true
        
        poi.altitude = 0
        
        poi.poiSize  = 2
        
        poi.poiCam   = poi_camera.POINTING_CAMERA
        
        poi.location = location
        
        _poisArray.add(poi)
        
        if Counter._counter % 2 == 0
        {
            poi.pathEAD = "data/Models/pois/knight/Dark_knight.EAD"
            poi.poiSize = 3
        }
        else
        {
            poi.pathEAD = "data/Models/pois/zombie/Zombie_head.EAD"
            poi.poiSize = 10
        }
        
        let uIdPOI = _my3dBrowser.add(poi)
        
        Counter._counter += 1
        
        if Counter._counter == 2
        {
            Counter._counter = 0
        }
        
        return uIdPOI == -1 ? false : true
    }
    
    func addPoi2D(location: CLLocation) -> Bool
    {
        struct Counter2D { static var _counter2D = 0 }
        
        let poi:Poi2D   = Poi2D()
        
        poi.altitude    = 0
        
        poi.setSelectedAlpha(1.0)
        
        poi.setNotSelectedAlpha(0.7)
        
        poi.poiSize     = 5
        
        poi.location    = location
        
        switch (Counter2D._counter2D)
        {
            case 0:
                poi.setTextureSelectionPath(NSString(string: "Viking_alpha.png") as String)
                poi.setTextureNonSelectionPath(NSString(string: "Transformer_alpha.png") as String)
                break
            case 1:
                poi.setTextureSelectionPath(NSString(string: "Zombie.png") as String)
                poi.setTextureNonSelectionPath(NSString(string: "Tv_alpha.png") as String)
                break
            default:
                break
        }
        
        _poisArray.add(poi)
        
        let uIdPOI = _my3dBrowser.add(poi)
        
        Counter2D._counter2D += 1
        
        if Counter2D._counter2D == 2
        {
            Counter2D._counter2D = 0
        }
        
        return uIdPOI == -1 ? false : true
    }
    
    func addPois(location: CLLocation) -> Bool
    {
        let _lat = location.coordinate.latitude
        let _lon = location.coordinate.longitude
        
        // Add a set of 2D and 3D POIs
        if addPoi(location: CLLocation(latitude: _lat - 1, longitude: _lon + 1)) && addPoi(location: CLLocation(latitude: _lat + 1, longitude: _lon + 1))
        {
            if addPoi2D(location: CLLocation(latitude: _lat + 1, longitude: _lon - 1)) && addPoi2D( location: CLLocation(latitude: _lat + 1, longitude: _lon - 2))
            {
                return true
            }
        }
        return false
    }
    
    
    // MARK: - Radar
    
    func setupRadar()
    {
        if _radarController == nil
        {
            _radarController = PoiRadarController(key: "67986b617b130cbf79bb0becd800c0a2a5fed428")
            
            assert(_radarController != nil, "Check API Key")
            
            var radarImage:UIImage
            
            if UI_USER_INTERFACE_IDIOM() == UIUserInterfaceIdiom.phone
            {
                radarImage = UIImage(named: "resources.bundle/iPhoneRadar.png")!
            }
            else
            {
                radarImage = UIImage(named: "resources.bundle/iPadRadar.png")!
            }
            
            _radarController.view.frame = CGRect(x: self.view.frame.origin.x + 10, y: self.view.frame.origin.y + 10, width: radarImage.size.width, height: radarImage.size.height)
            
            _radarController.radarImage = radarImage
            
            _radarController.poiImage           = UIImage(named: "resources.bundle/poi.png")
            _radarController.poiImageSelected   = UIImage(named: "resources.bundle/poiSelected.png")
            _radarController.northImage         = UIImage(named: "resources.bundle/northRadar.png")
            
            self.view.addSubview(_radarController.view)
        }
    }
    
    func removeRadar()
    {
        _radarController.view.removeFromSuperview()
        _radarController = nil
    }
    
    func rotateRadarView(degrees:Double)
    {
        _radarController.rotateRadarView(degrees)
    }
    
    func refreshPois()
    {
        let serialQueue = DispatchQueue(label: "com.aumentia.LockQueue")
        serialQueue.sync {
            
            self._radarController.poisArray = self._poisArray
            self._radarController.reset()
            self._radarController.refresh()
        }
    }
    
    func setUserLocationInRadar(userLocation:CLLocation)
    {
        _radarController.userLocation = userLocation
    }
    
    
    // MARK: - Render Delegate
    
    func poiClicked(_ poi: Poi!) {
        
        print("Poi \(poi.uId) clicked!!!")
        
        // Let's refresh the texture on click
        if let p = poi as? Poi2D
        {
            p.setTextureSelectionPath(NSString(string: "Knight_fullbody.png") as String)
            
            _my3dBrowser.refreshPoiTextures(p)
        }
        
        _currentSelectedPOI = poi
    }
    
    func poiSelected(_ poi: Poi!) {
        
        print("Poi \(poi.uId) selected!!!")
        
        _currentSelectedPOI = poi
    }
    
    func pois(inView poisArray: NSMutableArray!) {
        
        _radarController.selectedPoi = _currentSelectedPOI
        _radarController.refresh()
    }
}


// MARK: - Camera management
extension ViewController
{
    func initCapture()
    {
        // Init capture manager
        _captureManager = CaptureSessionManager()
        
        // Set video streaming quality
        _captureManager.captureSession.sessionPreset = AVCaptureSessionPresetPhoto
        
        _captureManager.outPutSetting = NSNumber(value: kCVPixelFormatType_32BGRA)
        
        _captureManager.addVideoInput(AVCaptureDevicePosition.back)
        _captureManager.addVideoOutput()
        _captureManager.addVideoPreviewLayer()
        
        let layerRect:CGRect = self.view.bounds
        
        _captureManager.previewLayer.isOpaque   = false
        _captureManager.previewLayer.bounds     = layerRect
        _captureManager.previewLayer.position   = CGPoint(x: layerRect.midX, y: layerRect.midY)
        
        // Create a view where we attach the AV Preview Layer
        _cameraView = UIView(frame: self.view.bounds)
        _cameraView.layer .addSublayer(_captureManager.previewLayer)
        
        // Add the view we just created as a subview to the View Controller's view
        self.view.addSubview(_cameraView)
        
        // Start
        DispatchQueue.main.async {
            
            self.startCaptureManager()
        }
    }
    
    func removeCapture()
    {
        _captureManager.captureSession.stopRunning()
        _cameraView.removeFromSuperview()
        _captureManager = nil
        _cameraView     = nil
    }
    
    func startCaptureManager()
    {
        autoreleasepool
            {
                _captureManager.captureSession.startRunning()
        }
    }
}


// MARK: - Motion Manager
extension ViewController
{
    func initMotionManager()
    {
        if _motionManager == nil
        {
            // Setup MotionManager
            _motionManager = CMMotionManager()
            
            _motionManager.showsDeviceMovementDisplay  = true
            
            _motionManager.accelerometerUpdateInterval = 1.0 / kAccelerometerFrequency
            _motionManager.deviceMotionUpdateInterval  = 1.0 / kAccelerometerFrequency
            
            _motionManager.startAccelerometerUpdates()
            _motionManager.startDeviceMotionUpdates(using: CMAttitudeReferenceFrame.xMagneticNorthZVertical)
        }
    }
    
    func removeMotionManager()
    {
        if _motionManager != nil
        {
            _locationManager.stopUpdatingHeading()
            _locationManager.stopUpdatingLocation()
            _locationManager = nil
            _lastLocation    = nil
        }
    }
}


// MARK: - Gesture Management
extension ViewController:UIGestureRecognizerDelegate
{
    func addGestureEvents()
    {
        // PAN
        if _panRecognizer == nil
        {
            _panRecognizer = UIPanGestureRecognizer(target: self, action: #selector(ViewController.move(recognizer:)))
            _panRecognizer.minimumNumberOfTouches = 1
            _panRecognizer.maximumNumberOfTouches = 1
            _panRecognizer.delegate               = self
            
            self.view.addGestureRecognizer(_panRecognizer)
        }
        
        // PINCH
        if _pinchRecognizer == nil
        {
            _pinchRecognizer = UIPinchGestureRecognizer(target: self, action: #selector(ViewController.zoomPinch(recognizer:)))
            
            self.view.addGestureRecognizer(_pinchRecognizer)
        }
        
        // ROTATION
        if _rotationRecognizer == nil
        {
            _rotationRecognizer = UIRotationGestureRecognizer(target: self, action: #selector(ViewController.spin(recognizer:)))
            
            self.view.addGestureRecognizer(_rotationRecognizer)
        }
    }
    
    @objc func spin(recognizer:UIRotationGestureRecognizer)
    {
        
    }
    
    @objc func zoomPinch(recognizer:UIPinchGestureRecognizer)
    {
        
    }
    
    @objc func move(recognizer:UIPanGestureRecognizer)
    {
        _my3dBrowser.spin3DPOI(recognizer)
    }
    
    func removeGestures()
    {
        if _panRecognizer != nil
        {
            self.view.removeGestureRecognizer(_panRecognizer)
        }
        if _pinchRecognizer != nil
        {
            self.view.removeGestureRecognizer(_pinchRecognizer)
        }
        if _rotationRecognizer != nil
        {
            self.view.removeGestureRecognizer(_rotationRecognizer)
        }
    }
}


// MARK: - Location Manager
extension ViewController: CLLocationManagerDelegate
{
    func setupLocationManager()
    {
        if _locationManager == nil
        {
            _locationManager                    = CLLocationManager()
            _locationManager.headingFilter      = kCLHeadingFilterNone
            _locationManager.desiredAccuracy    = kCLLocationAccuracyBest
            _locationManager.delegate           = self
            
            if _locationManager.responds(to: #selector(CLLocationManager.requestWhenInUseAuthorization))
            {
                _locationManager.requestWhenInUseAuthorization()
            }
            
            _locationManager.startUpdatingLocation()
            _locationManager.startUpdatingHeading()
        }
    }
    
    func removeLocationManager()
    {
        if _locationManager != nil
        {
            _locationManager.stopUpdatingHeading()
            _locationManager.stopUpdatingLocation()
        }
    }
    
    
    // MARK: - Location Delegate
    
    func locationManager(_ manager: CLLocationManager, didFailWithError error: Error) {
        
        print("Error: \(error)")
    }
    
    func locationManager(_ manager: CLLocationManager, didUpdateHeading newHeading: CLHeading) {
        
        _radarController.setHeading(newHeading)
    }
    
    func locationManager(_ manager: CLLocationManager, didUpdateLocations locations: [CLLocation]) {
        
        if let newLocation = locations.last {
            
            var oldLocation = locations.first
            
            if locations.count > 1 {
                
                oldLocation = locations[locations.count - 2]
            }
            
            let distanceInMeters:CLLocationDistance = newLocation.distance(from: oldLocation!)
            
            if _my3dBrowser != nil && distanceInMeters >= 0
            {
                if _lastLocation == nil
                {
                    // Add POIs
                    let _ = addPois(location: newLocation)
                    
                    refreshPois()
                }
                _lastLocation = newLocation
                
                print("Location updated: lon \(newLocation.coordinate.longitude) and lat \(newLocation.coordinate.latitude)")
                
                // ARBrowser
                _my3dBrowser.update(newLocation)
                
                // Radar
                setUserLocationInRadar(userLocation: newLocation)
            }
        }
    }
}

